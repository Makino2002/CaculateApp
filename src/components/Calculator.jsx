import React from "react";
import Button from "./button";
import Display from "./Display";
import History from "./History";
import useCalculator from "../hooks/useCalculator";

const Calculator = () => {
  const {
    display,
    isEditing,
    history,
    handleButtonClick,
    handleInputChange,
    setIsEditing,
  } = useCalculator();

  return (
    <div className="calculator">
      <Display
        value={display}
        onClick={() => setIsEditing(true)}
        onChange={handleInputChange}
        isEditing={isEditing}
        onBlur={() => setIsEditing(false)}
      />
      <div className="calculator-buttons">
        {[
          "AC",
          "+/-",
          "%",
          "÷",
          7,
          8,
          9,
          "✕",
          4,
          5,
          6,
          "-",
          1,
          2,
          3,
          "+",
          0,
          ".",
          "=",
        ].map((btn, i) => (
          <Button
            key={i}
            className={
              btn === 0
                ? "zero number"
                : typeof btn === "number"
                ? "number"
                : ["+", "-", "✕", "÷"].includes(btn)
                ? "operator"
                : btn === "="
                ? "equal operator"
                : btn === "."
                ? "number"
                : "other"
            }
            label={btn}
            onClick={handleButtonClick}
          />
        ))}
      </div>
      <History history={history} />
    </div>
  );
};

export default Calculator;
