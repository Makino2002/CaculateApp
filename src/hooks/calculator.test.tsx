import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Calculator from "../components/Calculator"; // Adjust the import to your file structure
import "@testing-library/jest-dom/extend-expect";
describe("Calculator", () => {
  test("UI Check", () => {
    const { container } = render(<Calculator />);
    expect(container).toMatchSnapshot();
  });

  test("Clear Functionality", () => {
    const { getByText, getByTestId } = render(<Calculator />);

    const display = getByTestId("calculator-display");
    const buttonAC = getByText("AC");
    const buttonOne = getByText("1");

    fireEvent.click(buttonOne);
    expect(display).toHaveTextContent("1");

    fireEvent.click(buttonAC);
    expect(display).toHaveTextContent("0");
  });

  test("Addition Operation", () => {
    const { getByText, getByTestId } = render(<Calculator />);

    const display = getByTestId("calculator-display");
    const buttonOne = getByText("1");
    const buttonTwo = getByText("2");
    const buttonAdd = getByText("+");
    const buttonEquals = getByText("=");

    fireEvent.click(buttonOne);
    fireEvent.click(buttonAdd);
    fireEvent.click(buttonTwo);
    fireEvent.click(buttonEquals);

    expect(display).toHaveTextContent("3");
  });

  test("Addition Without Operands", () => {
    const { getByText, getByTestId } = render(<Calculator />);

    const display = getByTestId("calculator-display");
    const buttonAdd = getByText("+");
    const buttonEquals = getByText("=");

    fireEvent.click(buttonAdd);
    fireEvent.click(buttonEquals);

    expect(display).toHaveTextContent("0");
  });

  test("Addition with Missing Second Operand", () => {
    const { getByText, getByTestId } = render(<Calculator />);

    const display = getByTestId("calculator-display");
    const buttonTwo = getByText("2");
    const buttonAdd = getByText("+");
    const buttonEquals = getByText("=");

    fireEvent.click(buttonTwo);
    fireEvent.click(buttonAdd);
    fireEvent.click(buttonEquals);

    expect(display).toHaveTextContent("4");
  });

  test("Subtraction Operation", () => {
    const { getByText, getByTestId } = render(<Calculator />);

    const display = getByTestId("calculator-display");
    const buttonThree = getByText("3");
    const buttonOne = getByText("1");
    const buttonSubtract = getByText("-");
    const buttonEquals = getByText("=");

    fireEvent.click(buttonThree);
    fireEvent.click(buttonSubtract);
    fireEvent.click(buttonOne);
    fireEvent.click(buttonEquals);

    expect(display).toHaveTextContent("2");
  });

  test("Division by Zero", () => {
    const { getByText, getByTestId } = render(<Calculator />);

    const display = getByTestId("calculator-display");
    const buttonFive = getByText("5");
    const buttonZero = getByText("0");
    const buttonDivide = getByText("รท");
    const buttonEquals = getByText("=");

    fireEvent.click(buttonFive);
    fireEvent.click(buttonDivide);
    fireEvent.click(buttonZero);
    fireEvent.click(buttonEquals);

    expect(display).toHaveTextContent("Error");
  });

  test("Multiple Decimal Points", () => {
    const { getByText, getByTestId } = render(<Calculator />);

    const display = getByTestId("calculator-display");
    const buttonOne = getByText("1");
    const buttonDot = getByText(".");

    fireEvent.click(buttonOne);
    fireEvent.click(buttonDot);
    fireEvent.click(buttonOne);
    fireEvent.click(buttonDot);

    expect(display).toHaveTextContent("1.1");
  });

  test("New Entry After Calculation", () => {
    const { getByText, getByTestId } = render(<Calculator />);

    const display = getByTestId("calculator-display");
    const buttonTwo = getByText("2");
    const buttonAdd = getByText("+");
    const buttonEquals = getByText("=");
    const buttonThree = getByText("3");

    fireEvent.click(buttonTwo);
    fireEvent.click(buttonAdd);
    fireEvent.click(buttonTwo);
    fireEvent.click(buttonEquals);
    expect(display).toHaveTextContent("4");

    fireEvent.click(buttonThree);
    expect(display).toHaveTextContent("3");
  });

  test("Toggle Sign Operation", () => {
    const { getByText, getByTestId } = render(<Calculator />);

    const display = getByTestId("calculator-display");
    const buttonOne = getByText("1");
    const buttonToggle = getByText("+/-");

    fireEvent.click(buttonOne);
    fireEvent.click(buttonToggle);
    expect(display).toHaveTextContent("-1");

    fireEvent.click(buttonToggle);
    expect(display).toHaveTextContent("1");
  });
});
