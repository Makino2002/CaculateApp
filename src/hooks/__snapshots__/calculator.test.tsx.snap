// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator UI Check 1`] = `
<div>
  <div
    class="calculator"
  >
    <input
      class="calculator-display "
      readonly=""
      type="text"
      value="0"
    />
    <div
      class="calculator-buttons"
    >
      <button
        class="button other"
      >
        AC
      </button>
      <button
        class="button other"
      >
        +/-
      </button>
      <button
        class="button other"
      >
        %
      </button>
      <button
        class="button operator"
      >
        ÷
      </button>
      <button
        class="button number"
      >
        7
      </button>
      <button
        class="button number"
      >
        8
      </button>
      <button
        class="button number"
      >
        9
      </button>
      <button
        class="button mul operator"
      >
        ✕
      </button>
      <button
        class="button number"
      >
        4
      </button>
      <button
        class="button number"
      >
        5
      </button>
      <button
        class="button number"
      >
        6
      </button>
      <button
        class="button operator"
      >
        -
      </button>
      <button
        class="button number"
      >
        1
      </button>
      <button
        class="button number"
      >
        2
      </button>
      <button
        class="button number"
      >
        3
      </button>
      <button
        class="button operator"
      >
        +
      </button>
      <button
        class="button zero number"
      >
        0
      </button>
      <button
        class="button number"
      >
        .
      </button>
      <button
        class="button equal operator"
      >
        =
      </button>
    </div>
    <div
      class="history"
    >
      <h3>
        History
      </h3>
      <ul />
    </div>
  </div>
</div>
`;
